File: trees/minimum_absolute_difference_in_bst.py

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from math import inf

class Solution:
    def getMinimumDifference(self, root: Optional[TreeNode]) -> int:

        def inorder_traversal(node: Optional[TreeNode]) -> None:
            if node is None:
                return

            inorder_traversal(node.left)
            
            nonlocal previous_value, min_difference
            min_difference = min(min_difference, node.val - previous_value)
            previous_value = node.val

            inorder_traversal(node.right)

        previous_value = -inf

        min_difference = inf

        inorder_traversal(root)

        return min_difference

File: trees/minimum_absolute_difference_in_bst.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer}
def get_minimum_difference(root)
    @prev_value = Float::INFINITY * -1
    @min_difference = Float::INFINITY

    def inorder_traversal(node)
        return if node.nil?

        inorder_traversal(node.left)

        @min_difference = [@min_difference, node.val - @prev_value].min
        @prev_value = node.val

        inorder_traversal(node.right)
    end

    inorder_traversal(root)
    @min_difference
end